cpp/include/raft/neighbors/detail/cagra/graph_core.cuh

```diff
@@ -538,309 +386,248 @@ void prune(raft::device_resources const& res,
-  for (int self = 0; self < num_gpus; self++) {
-    RAFT_CUDA_TRY(cudaSetDevice(self));
-    for (int peer = 0; peer < num_gpus; peer++) {
-      if (self == peer) { continue; }
-      RAFT_CUDA_TRY(cudaDeviceEnablePeerAccess(peer, 0));
-    }
-  }
-  RAFT_CUDA_TRY(cudaSetDevice(0));
-
-  uint32_t graph_chunk_size     = graph_size;
-  uint32_t*** d_input_graph_ptr = NULL;  // [...][num_gpus][graph_chunk_size, input_graph_degree]
-  graph_chunk_size              = (graph_size + num_gpus - 1) / num_gpus;
-  d_input_graph_ptr = mgpu_alloc<uint32_t>(num_gpus, graph_chunk_size, input_graph_degree);
+  auto input_graph_ptr               = knn_graph.data_handle();
+  auto output_graph_ptr              = new_graph.data_handle();
+  const IdxT graph_size              = new_graph.extent(0);
 
-  //
-  uint8_t* detour_count;  // [graph_size, input_graph_degree]
-  array_size   = sizeof(uint8_t) * graph_size * input_graph_degree;
-  detour_count = (uint8_t*)malloc(array_size);
-  memset(detour_count, 0xff, array_size);
+  auto pruned_graph = raft::make_host_matrix<IdxT, IdxT>(graph_size, output_graph_degree);
 
-  uint8_t*** d_detour_count = NULL;  // [...][num_gpus][graph_chunk_size, input_graph_degree]
-  d_detour_count            = mgpu_alloc<uint8_t>(num_gpus, graph_chunk_size, input_graph_degree);
-  mgpu_H2D<uint8_t>(
-    d_detour_count, detour_count, num_gpus, graph_size, graph_chunk_size, input_graph_degree);
-
-  //
-  uint32_t* num_no_detour_edges;  // [graph_size]
-  array_size          = sizeof(uint32_t) * graph_size;
-  num_no_detour_edges = (uint32_t*)malloc(array_size);
-  memset(num_no_detour_edges, 0, array_size);
-
-  uint32_t*** d_num_no_detour_edges = NULL;  // [...][num_gpus][graph_chunk_size]
-  d_num_no_detour_edges             = mgpu_alloc<uint32_t>(num_gpus, graph_chunk_size, 1);
-  mgpu_H2D<uint32_t>(
-    d_num_no_detour_edges, num_no_detour_edges, num_gpus, graph_size, graph_chunk_size, 1);
-
-  //
-  uint64_t** dev_stats  = NULL;  // [num_gpus][2]
-  uint64_t** host_stats = NULL;  // [num_gpus][2]
-  dev_stats             = (uint64_t**)malloc(sizeof(uint64_t*) * num_gpus);
-  host_stats            = (uint64_t**)malloc(sizeof(uint64_t*) * num_gpus);
-  array_size            = sizeof(uint64_t) * 2;
-  for (int i_gpu = 0; i_gpu < num_gpus; i_gpu++) {
-    RAFT_CUDA_TRY(cudaSetDevice(i_gpu));
-    RAFT_CUDA_TRY(cudaMalloc(&(dev_stats[i_gpu]), array_size));
-    host_stats[i_gpu] = (uint64_t*)malloc(array_size);
-  }
-  RAFT_CUDA_TRY(cudaSetDevice(0));
+  {
+    //
+    // Prune kNN graph
+    //
+    auto d_input_graph = raft::make_device_matrix<IdxT, IdxT>(res, graph_size, input_graph_degree);
+
+    auto detour_count = raft::make_host_matrix<uint8_t, IdxT>(graph_size, input_graph_degree);
+    auto d_detour_count =
+      raft::make_device_matrix<uint8_t, IdxT>(res, graph_size, input_graph_degree);
+    RAFT_CUDA_TRY(cudaMemset(
+      d_detour_count.data_handle(), 0xff, graph_size * input_graph_degree * sizeof(uint8_t)));
```

Comment:
While practically it does not really matter, it is more consistent if we use `cudaMemsetAsync`
```suggestion
    RAFT_CUDA_TRY(cudaMemsetAsync(
      d_detour_count.data_handle(), 0xff, graph_size * input_graph_degree * sizeof(uint8_t), res.get_stream()));
```